File: code/ch02/2.4.1.fourSumCount.java
import java.util.*;

public class Solution {
  public int fourSumCount(int[] A,
      int[] B, int[] C, int[] D) {
    Map<Integer, Integer> mapper = new HashMap<Integer, Integer>();
    int res = 0;
    for (int i = 0; i < A.length; i++)
      for (int j = 0; j < B.length; j++) {
        if (mapper.get(A[i] + B[j]) != null)
          mapper.put(A[i] + B[j], mapper.get(A[i] + B[j]) + 1);
        else
          mapper.put(A[i] + B[j], 1);
      }
    for (int i = 0; i < C.length; i++)
      for (int j = 0; j < D.length; j++)
        if (mapper.get(-1 * (C[i] + D[j])) != null)
          res += mapper.get(-1 * (C[i] + D[j]));
    return res;
  }
}

File: code/ch02/2.4.1.fourSumCount.py
class Solution:
    def fourSumCount(
        self, A: List[int], B: List[int], C: List[int], D: List[int]
    ) -> int:
        mapper = {}
        res = 0
        for i in A:
            for j in B:
                mapper[i + j] = mapper.get(i + j, 0) + 1

        for i in C:
            for j in D:
                res += mapper.get(-1 * (i + j), 0)
        return res
